### PR TITLE
Moving bbox-related functionality to bbox module

### PR DESCRIPTION
**Pull request by dkoguciuk**
### Description
The fix to #688, including:
1) Creation of a new module named `bbox` under `kornia.geometry`
2) The funcions are renamed to `*_2d` and `*_3d`.

Maybe we could merge bbox-related functionality for 2D and 3D into one dimensionality-agnostic version, where depending on the shape of the input bbox we would call 2d or 3d version of the function (in the next PR?).

### Status
**Ready**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
- [x] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [x] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>

  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?

</details>

### CODE REVIEW

**Comment by shijianjian** [comment]
Does it have backward compatibility?

**Review by edgarriba — COMMENTED** [review]
make sure of backward compatibility - update also the docs

**Comment by dkoguciuk** [comment]
Hi @shijianjian ,

no, it breaks backward compatibility in two ways:
1. Methods are moved to different place (`kornia.geometry.bbox` instead of `kornia.geometry.transform.crop.crop2d`)
2. Methods are renamed to follow the same pattern (i.e `infer_bbox_shape_2d` instead of `infer_box_shape`)
3. I can add those methods back with a deprecation warning, but we will have to keep it in mind to delete it someday.

Of course all of above is discussable :smiley:

File: kornia/augmentation/augmentation.py [hunk 1]
```diff
@@ -22,8 +22,8 @@
 from kornia.filters import box_blur, gaussian_blur2d, motion_blur
 from kornia.geometry import (
     affine,
-    bbox_generator,
-    bbox_to_mask,
+    bbox_generator_2d,
```
[the remark this answers is not in the document]
Comment (dkoguciuk) [reply kornia/augmentation/augmentation.py:25]: @edgarriba , is it really a good approach? `bbox_generator` will be 2d as default and `bbox_generator3d` will be for 3d -- is it really what we want?

The second thing is that some functions have to be renamed (i.e. `infer_bbox_shape` instead of `infer_box_shape`).

File: kornia/geometry/bbox.py [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+        return validate_bbox_2d(boxes)
+
+    # 3D boxes
+    elif boxes.shape[-1] == 3:
+        return validate_bbox_3d(boxes)
+
+    return True
+
+
+def infer_bbox_shape_2d(boxes: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
```
[the remark this answers is not in the document]
Comment (dkoguciuk) [reply kornia/geometry/bbox.py:103]: This is the tricky part. I've prepared `validate_bbox` as an example of what I want to do in the future. Adding generic `infer_bbox_shape` doesn't work when wrapping with `jit`, because it returns `Tuple[torch.Tensor, torch.Tensor]` in 2d case and `Tuple[torch.Tensor, torch.Tensor, torch.Tensor]` in 3d case and ASAIK `jit` doesn't support Union so far. I am planning to add a generic function for every bbox-related method, but it will take more time, because I don't know `jit`. Do you want me to do in in this PR?

**Comment by edgarriba** [comment]
@dkoguciuk we'll have soon the api discussion of 2d/3d methods

**Comment by dkoguciuk** [comment]
@edgarriba , should I keep the old naming then? `bbox_generator` for 2d and `bbox_generator3d` for 3d?

File: kornia/augmentation/container/augment.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 
 from kornia.augmentation.base import _AugmentationBase, GeometricAugmentationBase2D, IntensityAugmentationBase2D
 from kornia.constants import DataKey
-from kornia.geometry import transform_boxes, transform_points
+from kornia.geometry import transform_bbox_2d, transform_points
```
[the remark this answers is not in the document]
Comment (dkoguciuk) [reply kornia/augmentation/container/augment.py]: Done!

File: kornia/geometry/bbox.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dkoguciuk) [reply kornia/geometry/bbox.py:103]: As a dirty workaround we can for now add `@torch.jit.ignore` with `infer_bbox_shape` but allow to wrap it around `infer_bbox_shape_2d` and `infer_bbox_shape_3d`.

File: kornia/geometry/linalg.py [hunk 4]
```diff
@@ -205,45 +204,6 @@ def transform_points(trans_01: torch.Tensor, points_1: torch.Tensor) -> torch.Te
     return points_0
 
 
-def transform_boxes(trans_mat: torch.Tensor, boxes: torch.Tensor, mode: str = "xyxy") -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (dkoguciuk) [reply kornia/geometry/linalg.py]: Done!

File: kornia/geometry/bbox.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dkoguciuk) [reply kornia/geometry/bbox.py:100]: Ok, generic `infer_bbox_shape` is added, I'll add generic `bbox_to_mask` and `bbox_generator` soon.

File: kornia/geometry/linalg.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dkoguciuk) [reply kornia/geometry/linalg.py]: Done!

File: kornia/geometry/bbox.py [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+
+    bot = torch.index_select(boxes, 1, torch.tensor([2, 3, 6, 7], device=boxes.device, dtype=torch.long))[:, :, 1]
+    upper = torch.index_select(boxes, 1, torch.tensor([0, 1, 4, 5], device=boxes.device, dtype=torch.long))[:, :, 1]
+    heights = (bot - upper + 1)[:, 0]
+
+    depths = (boxes[:, 4:, 2] - boxes[:, :4, 2] + 1)[:, 0]
+    return depths, heights, widths
+
+
+def bbox_to_mask_2d(boxes: torch.Tensor, width: int, height: int) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (dkoguciuk) [reply kornia/geometry/bbox.py:223]: To have a generic `bbox_to_mask` we would have to change the API. Currently we have:

1. `def bbox_to_mask_2d(boxes: torch.Tensor, width: int, height: int) -> torch.Tensor:` for 2d case
2. `def bbox_to_mask_3d(boxes: torch.Tensor, size: Tuple[int, int, int]) -> torch.Tensor` for 3d case

So we either do the merging as it is now, without generic bbox_to_mask function, or we do API change, i.e. to:

1. `def bbox_to_mask_2d(boxes: torch.Tensor, size: Tuple[int, int]) -> torch.Tensor:` for 2d case
2. `def bbox_to_mask_3d(boxes: torch.Tensor, size: Tuple[int, int, int]) -> torch.Tensor` for 3d case
3. `def bbox_to_mask(boxes: torch.Tensor, size: Union[Tuple[int, int], Tuple[int, int, int]]) -> torch.Tensor:` for generic case

**Comment by shijianjian** [comment]
> @edgarriba , should I keep the old naming then? `bbox_generator` for 2d and `bbox_generator3d` for 3d?

Personally, I think we should keep the original naming to be consistent with other modules.

**Comment by dkoguciuk** [comment]
Sure, how do you propose to name generic functions then? @shijianjian

1. `bbox_generator` for 2d 
2. `bbox_generator3d` for 3d
3. `here_the_generic version`

**Comment by shijianjian** [comment]
> Sure, how do you propose to name generic functions then? @shijianjian
> 
> 
> 
> 1. `bbox_generator` for 2d 
> 
> 2. `bbox_generator3d` for 3d
> 
> 3. `here_the_generic version`

Not sure why do we need a bbox more than 3D. But how about bbox_generator_nd ?

**Comment by dkoguciuk** [comment]
Hi @shijianjian ,

I am not sure if we need bboxes with more than 3 dimensions. I thought about preparing one function for both 2d and 3d bounding boxes. Does it make any sense?

Best,
Daniel

**Comment by shijianjian** [comment]
> I am not sure if we need bboxes with more than 3 dimensions. I thought about preparing one function for both 2d and 3d bounding boxes. Does it make any sense?

I do not think so. I think we introduce a new function as if to provide a new feature or as a syntax sugar. 

So far, I cannot see the point to provide a function so that the user will not need to think about if they are dealing with 2D or 3D data. The user shall aware what they are dealing with in the first place.

I would suggest to not include the generic bbox feature in this PR, and potentially to discuss in a separate issue firstly.

**Comment by dkoguciuk** [comment]
@shijianjian, yes, I think you're right. At first, I thought that generic functions would be cleaner and have less duplicated code, but it is probably overcomplicated. @edgarriba , do you agree?

File: kornia/geometry/bbox.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/geometry/bbox.py:223]: I like the generic - however Union implies issues with torchscript for now.

**Comment by edgarriba** [comment]
Then let's try to keep if separated if the internals are over-complicated. I'd prefer readability and easy comprehension to later keep it easy to maintain.

**Comment by dkoguciuk** [comment]
@edgarriba , @shijianjian , can you please take a look now?

**Review by edgarriba — APPROVED** [review]
LGTM. Don't forget to add to the docs before merging

**Comment by dkoguciuk** [comment]
@shijianjian , I've changed naming from `*_3d` to `*3d` to be more aligned with the previous naming conventions. Can you take a look?

File: kornia/augmentation/container/augment.py [hunk 6]
```diff
@@ -7,7 +7,8 @@
 
 from kornia.augmentation.base import _AugmentationBase, GeometricAugmentationBase2D, IntensityAugmentationBase2D
 from kornia.constants import DataKey
-from kornia.geometry import transform_boxes, transform_points
+from kornia.geometry.bbox import transform_bbox
```
Comment (shijianjian) [inline kornia/augmentation/container/augment.py:10]: Why not plural? We used transform_points as well.

**Review by shijianjian — APPROVED** [review]
LGTM. Just one naming thing. 

BTW, regarding the generalized API, I think it might be good to have a suite of API for non-rect box. On top of that, we have a regular shaped bbox like this. 

One motivation can be some researches are using rotated bbox. Cannot recall the paper name. XD.

**Comment by dkoguciuk** [comment]
Hi @shijianjian ,

fair point! Please help me to design consistent naming:
1. old `validate_bboxes`  -> new `validate_bboxes`
2. old `infer_box_shape` -> new `infer_bboxes_shape`
3. old `bbox_to_mask` -> new `bboxes_to_masks`
4. old `bbox_generator` -> new `bbox_generator`
5. old `transform_bbox` -> new `transform_bboxes`

Can be, however I liked singular version more TBH. How do you think @shijianjian ?

File: kornia/augmentation/container/augment.py (hunk 6 above)
Comment (dkoguciuk) [reply kornia/augmentation/container/augment.py:10]: Discussed on slack.

File: kornia/augmentation/augmentation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dkoguciuk) [reply kornia/augmentation/augmentation.py]: Discussed this on slack.

File: kornia/geometry/bbox.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dkoguciuk) [reply kornia/geometry/bbox.py]: Discussed this on slack.

File: kornia/geometry/bbox.py [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+        m_out = torch.ones_like(m)
+        m_out = m_out * (m == 1).all(dim=2, keepdim=True).all(dim=1, keepdim=True)
+        m_out = m_out * (m == 1).all(dim=3, keepdim=True).all(dim=1, keepdim=True)
+        m_out = m_out * (m == 1).all(dim=2, keepdim=True).all(dim=3, keepdim=True)
+        mask_out.append(m_out)
+
+    return torch.stack(mask_out, dim=0).float()
+
+
+def bbox_generator_2d(
```
[the remark this answers is not in the document]
Comment (dkoguciuk) [reply kornia/geometry/bbox.py]: For now, we won't as discussed.

File: kornia/geometry/bbox.py [hunk 8]
```diff
@@ -0,0 +1,500 @@
...
+
+    # back
+    bbox_back = bbox.clone()
+    bbox_back[:, :, -1] += depth.unsqueeze(dim=1).repeat(1, 4)
+    bbox = torch.cat([bbox, bbox_back], dim=1)
+
+    return bbox
+
+
+def transform_bbox_2d(trans_mat: torch.Tensor, boxes: torch.Tensor, mode: str = "xyxy") -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (dkoguciuk) [reply kornia/geometry/bbox.py]: For now, we won't as discussed.

File: docs/source/geometry.bbox.rst [hunk 9]
```diff
@@ -0,0 +1,7 @@
+kornia.geometry.homography
```
Comment (edgarriba) [inline docs/source/geometry.bbox.rst:1]: @dkoguciuk error here

Comment (edgarriba) [reply docs/source/geometry.bbox.rst:1]: And old references should be removed

**Comment by shijianjian** [comment]
@dkoguciuk Hi, when I run tests like "pytest test/augmentation/test_containers.py", I got the deprecation warning message. Can you fix that as well that we should not receive deprecation warnings inside our lib.

**Comment by dkoguciuk** [comment]
Hi @shijianjian ,  yes, sure, sorry for the mistake. It's fixed in #1141, you can take a look and approve it :smiley:
